### PR TITLE
Make admins adhere to branch protection rules

### LINKED ISSUE #17131: Option to turn off ability for administrator to merge pull request without getting approvals granted
### Feature Description

Say I am one of the 'owners' of a repo and I've set branch protection on the `main` branch. Specifically, I've set required approvals to 3. 

This works fine for the most part, but on the pull request it shows 'As an administrator, you may still merge this pull request.':

image

So I could click the red 'Rebase and Merge' button and bypass the required approvals.

A feature I would really like would be a way to disable this ability. So that, even as an owner, I don't get this option to merge the PR without approvals. 

Now I understand that, as owner of the repo, I have ultimate control of the settings. If I wanted to, I could go into settings and remove the need for required approvals. But that is quite an involved process which under normal circumstances I wouldn't do.

However, I want to avoid the temptation for an owner to quickly merge a PR without getting approvals granted first.

If I could just disable this merge button from showing, until approvals have been granted, then that would improve usability. 

### Screenshots

_No response_

### PR DESCRIPTION
**Pull request by enko**
This introduces a new flag `BlockAdminMergeOverride` on the branch protection rules that prevents admins/repo owners from bypassing branch protection rules and merging without approvals or failing status checks.

Fixes #17131

### CODE REVIEW

**Comment by lunny** [comment]
Wouldn't the admin can change the protection rules?

File: templates/repo/issue/view_content/pull.tmpl [hunk 1]
```diff
@@ -164,7 +164,7 @@
 				{{$notAllOverridableChecksOk := or .IsBlockedByApprovals .IsBlockedByRejection .IsBlockedByOfficialReviewRequests .IsBlockedByOutdatedBranch .IsBlockedByChangedProtectedFiles (and .EnableStatusCheck (not .RequiredStatusCheckState.IsSuccess))}}
 
 				{{/* admin can merge without checks, writer can merge when checks succeed */}}
```
Comment (wxiaoguang) [inline templates/repo/issue/view_content/pull.tmpl:166]:
```suggestion
				{{/* admin can merge without checks, writer can merge when checks succeed, BlockAdminMergeOverride only hides the button on web UI but doesn't stop the admin from using API or web request to force merging */}}
```

I think this change means "UI-only hidden", but not completely prevent admin users from merging.

**Comment by enko** [comment]
> Wouldn't the admin can change the protection rules?

Yes, but that would leave a paper trail. 

It is the same on GitHub.

File: templates/repo/issue/view_content/pull.tmpl (hunk 1 above)
Comment (enko) [reply templates/repo/issue/view_content/pull.tmpl:166]: Thanks for the pointer. I'll look into changing the API endpoints for merging pull requests.

**Comment by lunny** [comment]
> > Wouldn't the admin can change the protection rules?
> 
> Yes, but that would leave a paper trail.
> 
> It is the same on GitHub.

I know Github's behaviour but I suspect it's not a better solution than having a configuration item in app.ini

**Comment by enko** [comment]
> > I know Github's behaviour but I suspect it's not a better solution than having a configuration item in app.ini

I see why you would want to have it in your `app.ini` to have this setting behave very strict. My point would be that it is primarily about compliance and accidental merge. If an owner/admin changes the setting, he has to actively do it and if Gitea would have an audit log, it would leave a paper trail.

> But why a "global" config item? For large instance, different organizations have different requirements.

Yes, different projects have different needs and sometimes even different branches have different needs.

> […] Gitea does need a well-designed config system, global -> org-level -> repo-level.

That is out of scope.

**Comment by enko** [comment]
@wxiaoguang I added the service changes and also added a test case.

**Comment by wxiaoguang** [comment]
Thank you for the update. There are still some things left:

1. There are some lint errors.
1. Some structs in "modules/structs/repo_branch.go" are changed, but I do not see how they are used in code. If I read correctly, only the one in "ProtectBranchForm" is really used in code.
1. Database change needs a database migration.

**Comment by enko** [comment]
@wxiaoguang Thanks for your input!

> Thank you for the update. There are still some things left:
> 
>     1. There are some lint errors.

Fixed that, sorry for that.

>     2. Some structs in "modules/structs/repo_branch.go" are changed, but I do not see how they are used in code. If I read correctly, only the one in "ProtectBranchForm" is really used in code.

I checked, and I noticed I missed the API did not use the new field. I added that.

>     3. Database change needs a database migration.

Added a migration, I hope I did that right?

File: services/pull/check.go [hunk 2]
```diff
@@ -118,13 +118,26 @@ func CheckPullMergeable(stdCtx context.Context, doer *user_model.User, perm *acc
 				err = nil
 			}
 
-			// * if the doer is admin, they could skip the branch protection check
+			// * if the doer is admin, they could sometimes skip the branch protection check
 			if adminSkipProtectionCheck {
-				if isRepoAdmin, errCheckAdmin := access_model.IsUserRepoAdmin(ctx, pr.BaseRepo, doer); errCheckAdmin != nil {
+				pb, pbErr := git_model.GetFirstMatchProtectedBranchRule(ctx, pr.BaseRepoID, pr.BaseBranch)
+				if pbErr != nil {
+					return fmt.Errorf("LoadProtectedBranch: %v", err)
```
Comment (wxiaoguang) [inline services/pull/check.go]: It seems not right. It should use `pbErr` but not `err`, and `Errorf` should use `%w` to wrap the error.

I will make some changes.

**Comment by wxiaoguang** [comment]
Some changes in 71b5edd500b282d312ca068679d7f41b934d86a4:

1. Make the database migration work `models/migrations/migrations.go`)
1. Add some spaces in `options/locale/locale_en-US.ini`
1. Use `Errorf(%w)` to wrap the correct errors
1. Simplify some `if` blocks
1. Rename the argument `adminSkipProtectionCheck` to `adminForceMerge` to match the callers `form.ForceMerge`.

File: templates/repo/issue/view_content/pull.tmpl [hunk 3]
```diff
@@ -164,7 +164,7 @@
...
-				{{$canMergeNow := and (or $.IsRepoAdmin (not $notAllOverridableChecksOk)) (or (not .AllowMerge) (not .RequireSigned) .WillSign)}}
+				{{$canMergeNow := and (or (and (not $.BlockAdminMergeOverride) $.IsRepoAdmin) (not $notAllOverridableChecksOk)) (or (not .AllowMerge) (not .RequireSigned) .WillSign)}}
```
Comment (wxiaoguang) [inline templates/repo/issue/view_content/pull.tmpl]: ~~It seems that the red "merge" button is still on the UI?~~

<details>

image


</details>

----


Updte: hmm .... it is for "auto merge", not "force merge", the color is strange (another problem)

**Review by wxiaoguang — APPROVED** [review]

File: models/migrations/migrations.go [hunk 4]
```diff
@@ -603,6 +603,8 @@ var migrations = []Migration{
 	NewMigration("Add index for release sha1", v1_23.AddIndexForReleaseSha1),
 	// v305 -> v306
 	NewMigration("Add Repository Licenses", v1_23.AddRepositoryLicenses),
+	// v306 -> v307
+	NewMigration("Add Repository Licenses", v1_23.AddBlockAdminMergeOverrideBranchProtection),
```
Comment (lalo-mx) [inline models/migrations/migrations.go]: Migration title is repeated from last migration

Comment (wxiaoguang) [reply models/migrations/migrations.go]:
```suggestion
	NewMigration("Add BlockAdminMergeOverride to ProtectedBranch", v1_23.AddBlockAdminMergeOverrideBranchProtection),
```

Oooooooooops, I must be not awaken .....

**Comment by lunny** [comment]
图片
Scheduled merge should also be checked from template.

**Comment by wxiaoguang** [comment]
> Scheduled merge should also be checked from template.

Why?


image

**Comment by lunny** [comment]
I mean auto-merge means should be hidden for the administrator from the UI.

**Comment by wxiaoguang** [comment]
> I mean auto-merge means should be hidden for the administrator from the UI.

Why they should hide? Auto merge only merge when checks pass.

**Comment by lunny** [comment]
> > I mean auto-merge means should be hidden for the administrator from the UI.
> 
> Why they should hide? Auto merge only merge when checks pass.

Yes, but the permissions that can merge the pull request are checked when creating the schedule.

**Comment by wxiaoguang** [comment]
> > > I mean auto-merge means should be hidden for the administrator from the UI.
> > 
> > 
> > Why they should hide? Auto merge only merge when checks pass.
> 
> Yes, but the permissions that can merge the pull request are checked when creating the schedule.

I do not understand what you mean. Show a real case, what's wrong would happen.

**Review by lunny — APPROVED** [review]

**Comment by lunny** [comment]
> @wxiaoguang @lunny Thank you for your support in bringing this feature to fruition.
> 
> Is there anything for me to do?

No action is needed unless additional maintainers submit new review requests. Otherwise, this pull request will be merged in a few days.

**Pull request merged**